L13-python-interceptors.md

```diff
@@ -0,0 +1,274 @@
+
+```Python
+class HandlerCallDetails(six.with_metaclass(abc.ABCMeta)):
+    """Describes an RPC that has just arrived for service.
+
+    Attributes:
+      method: The method name of the RPC.
+      invocation_metadata: The metadata sent by the client.
+    """
+```
+
+The `Continuation` in the server side is thinner than the client side. It is a function that takes a `grpc.HandlerCallDetails` and proceeds to invoke the next interceptor in the chain, if any, or the RPC handler lookup logic, with the call details, passed as an argument and returns a `grpc.RpcMethodHandler` instance if the RPC is considered serviced, or None otherwise.
+
+The server-side interceptor interface contains one abstract member function that takes in `Continuation`, and `grpc.HandlerCallDetails`, then returns a `grpc.RpcMethodHandler`.
+
+```Python
+class ServerInterceptor(six.with_metaclass(abc.ABCMeta)):
+    @abc.abstractmethod
+    def intercept_service(self, continuation, handler_call_details): pass
+```
+
+To apply interception, the [server construction function](https://grpc.io/grpc/python/grpc.html#grpc.server) takes in an `interceptors` parameter that accepts optional list of `grpc.ServerInterceptor` instance. The usage snippet will look like:
+
+```Python
+class ExampleInterceptor(grpc.ServerInterceptor):
+    def intercept_service(self, continuation, handler_call_details):
+        ...
+
+server_interceptor_1 = ExampleInterceptor(...)
+server = grpc.server(..., interceptors=[server_interceptor_1])
+```
+
+### Server-Side Implementation
+
+In the implementation, there is a new concept named "Service Pipeline". It assembles the server side interceptors and is responsible for generating the final `grpc.RpcMethodHandler` to be invoked for incoming requests. The generation process will happen every time the server receives an RPC, the execution flow will look like:
+
+```
+Server Receives a Request ->
+    Interceptor A Start ->
+        Interceptor B Start ->
+            Interceptor C Start ->
+                The Original Handler
+            Interceptor C Returns Updated Handler C ->
+        Interceptor B Returns Updated Handler B ->
+    Interceptor A Returns Updated Handler A ->
+
+    Invoke the Updated Handler A with the Request ->
+    Updated Handler A Returns Response ->
+Server Replies
+```
+
+The most significant difference compared to the client side is that the interception happens ahead of the actual invocation of application behavior. So the interceptor function will be able to access neither the request nor servicer context.
+
+## Rationale
+
+### The Benefit of Introducing "Continuation"
+
+Quote from mehrdada@ [design doc](https://github.com/grpc/proposal/blob/361846400834164d21c12ca6cf60e8cbf501d6cd/L13-Python-Interceptors.md):
+> All of the methods are passed a callable value `Continuation`, whose
```

Comment:
This probably should not be a "quote" since the actual document uses `invoker`....

Reply:
The "invoker" and "continuation" are the same concept, I updated the line to:

"Quote from mehrdada@ design doc, the `invoker` below is the same concept of `continuation` function:"

Alternatively, we can remove the whole section, since the behaviors described here are elaborated above in the Client-Side Implementation section.